docs/architecture/architecture.md

```diff
@@ -0,0 +1,378 @@
+
+#### Runtime
+
+`cc-runtime` is an OCI compatible container runtime and is responsible for handling all
+commands specified by [the OCI runtime specification](https://github.com/opencontainers/runtime-spec)
+and launching `cc-shim` instances.
+
+Here we will describe how `cc-runtime` handles the most important OCI commands.
+
+##### `create`
+
+When handling the OCI `create` command, `cc-runtime` goes through the following steps:
+
+1. Create the container namespaces (Only the network and mount namespaces are currently supported),
+according to the container OCI configuration file.
+2. Spawn the `cc-shim` process and have it wait on a couple of temporary pipes for:
+   * A `cc-proxy` created file descriptor (one end of a socketpair) for the shim to connect to.
+   * The container `agent` sequence numbers for at most 2 I/O streams (One for `stdout` and `stdin`, one for `stderr`).
+   The `agent` uses those sequence numbers to multiplex all streams for all processes through one serial interface (The
+   virtio I/O serial one).
+3. Run all the [OCI hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md#hooks) in the container namespaces,
+as described by the OCI container configuration file.
+4. **fixme** [Set up the container networking](https://github.com/clearcontainers/runtime/blob/master/documentation/architecture.md#networking).
+This must happen after all hooks are done as one of them is potentially setting
+the container networking namespace up.
+5. Create the virtual machine running the container process. The VM `systemd` instance will spawn the `agent` daemon.
+6. Wait for the `agent` to signal that it is ready.
+7. Send the pod creation command to the `agent`. The `agent` pod is the container process sandbox.
+8. Send the allocateIO command to the proxy, for getting the `agent` I/O sequence numbers described in step 2.
+9. Pass the `cc-proxy` socketpair file descriptor, and the I/O sequence numbers to the listening cc-shim process through the dedicated pipes.
+10. The `cc-shim` instance is put into a stopped state to prevent it from doing I/O before the container is started.
+
+![Docker create](create.png)
+
+At that point, the container sandbox is created in the virtual machine and `cc-shim` is stopped on the host.
+However the container process itself is not yet running as one needs to call `docker start` to actually start it.
+
+##### `start`
+
+On namespace containers `start` launches a traditional Linux container process in its own set of namespaces.
+With Clear Containers, the main task of `cc-runtime` is to create and start a container within the
+pod that got created during the `create` step. In practice, this means `cc-runtime` follows
+these steps:
+
+1. `cc-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know which pod
+we want to use to create and start the new container.
+2. `cc-runtime` sends an agent `NEWCONTAINER` command to create and start a new container in
```

Comment:
Before that step, we also send the `STARTPOD` command, and then one `NEWCONTAINER` per container on the pod.